core/base/composition.cpp

```diff
@@ -33,63 +33,89 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/base/composition.hpp>
 
 
-#include <ginkgo/core/matrix/dense.hpp>
+#include <algorithm>
 
 
-namespace gko {
-namespace {
+#include <ginkgo/core/matrix/dense.hpp>
 
 
-template <typename ValueType, typename OpIterator, typename VecIterator>
-inline void allocate_vectors(OpIterator begin, OpIterator end, VecIterator res)
-{
-    for (auto it = begin; it != end; ++it) {
-        if (*res == nullptr || (*res)->get_size()[0] != (*it)->get_size()[0]) {
-            *res = matrix::Dense<ValueType>::create(
-                (*it)->get_executor(), gko::dim<2>{(*it)->get_size()[0], 1});
-        }
-        ++res;
-    }
-}
+namespace gko {
 
 
-inline const LinOp *apply_inner_operators(
+template <typename ValueType>
+std::unique_ptr<LinOp> apply_inner_operators(
     const std::vector<std::shared_ptr<const LinOp>> &operators,
-    const std::vector<std::unique_ptr<LinOp>> &intermediate, const LinOp *rhs)
+    Array<ValueType> &storage, const LinOp *rhs)
 {
-    for (auto i = operators.size() - 1; i > 0u; --i) {
-        auto solution = lend(intermediate[i - 1]);
-        operators[i]->apply(rhs, solution);
-        rhs = solution;
+    using Dense = matrix::Dense<ValueType>;
+    // determine amount of necessary storage:
+    // maximum sum of two subsequent intermediate vectors
+    // (and the out dimension of the last op if we only have one operator)
+    auto num_rhs = rhs->get_size()[1];
+    auto max_intermediate_size = std::accumulate(
+        begin(operators) + 1, end(operators) - 1,
+        operators.back()->get_size()[0],
+        [](size_type acc, std::shared_ptr<const LinOp> op) {
+            return std::max(acc, op->get_size()[0] + op->get_size()[1]);
+        });
+    auto storage_size = max_intermediate_size * num_rhs;
```

Comment:
`max_intermediate_size` can be calculated in Composition creation.

Reply:
I tried, but due to the variadic constructor, it is impossible to determine if the constructor call is the final one, thus making it impossible to "ignore" the first operator. So I decided to do it here, since the overhead is minimal and we don't need to duplicate any code.

Reply:
I see.